generator/altitude_generator.cpp

```diff
@@ -0,0 +1,247 @@
+#include "base/string_utils.hpp"
+
+#include "defines.hpp"
+
+#include "3party/succinct/elias_fano.hpp"
+#include "3party/succinct/mapper.hpp"
+#include "3party/succinct/rs_bit_vector.hpp"
+
+#include "std/algorithm.hpp"
+#include "std/type_traits.hpp"
+#include "std/utility.hpp"
+#include "std/vector.hpp"
+
+using namespace feature;
+
+namespace
+{
+TAltitudeSectionVersion constexpr kAltitudeSectionVersion = 1;
+
+class Processor
+{
+public:
+  using TFeatureAltitude = pair<uint32_t, Altitude>;
+  using TFeatureAltitudes = vector<TFeatureAltitude>;
+
+  Processor(string const & srtmPath) : m_srtmManager(srtmPath), m_minAltitude(kInvalidAltitude) {}
+
+  TFeatureAltitudes const & GetFeatureAltitudes() const { return m_featureAltitudes; }
+
+  vector<bool> const & GetAltitudeAvailability() const { return m_altitudeAvailability; }
+
+  TAltitude GetMinAltitude() const { return m_minAltitude; }
+
+  void operator()(FeatureType const & f, uint32_t const & id)
+  {
+    if (id != m_altitudeAvailability.size())
+    {
+      LOG(LERROR, ("There's a gap in feature id order."));
+      return;
+    }
+
+    bool hasAltitude = false;
+    do
+    {
+      if (!routing::IsRoad(feature::TypesHolder(f)))
+        break;
+
+      f.ParseGeometry(FeatureType::BEST_GEOMETRY);
+      size_t const pointsCount = f.GetPointsCount();
+      if (pointsCount == 0)
+        break;
+
+      TAltitudes altitudes;
+      bool allPointsValidAltFlag = true;
+      TAltitude minFeatureAltitude = kInvalidAltitude;
+      for (size_t i = 0; i < pointsCount; ++i)
+      {
+        TAltitude const a = m_srtmManager.GetHeight(MercatorBounds::ToLatLon(f.GetPoint(i)));
+        if (a == kInvalidAltitude)
+        {
+          allPointsValidAltFlag = false;
+          break;
+        }
+
+        if (minFeatureAltitude == kInvalidAltitude)
+          minFeatureAltitude = a;
+        else
+          minFeatureAltitude = min(minFeatureAltitude, a);
+
+        altitudes.push_back(a);
+      }
+      if (!allPointsValidAltFlag)
+        break;
+
+      hasAltitude = true;
+      m_featureAltitudes.push_back(make_pair(id, Altitude(move(altitudes))));
+
+      if (m_minAltitude == kInvalidAltitude)
+        m_minAltitude = minFeatureAltitude;
+      else
+        m_minAltitude = min(minFeatureAltitude, m_minAltitude);
+    }
+    while(false);
+
+    m_altitudeAvailability.push_back(hasAltitude);
+  }
+
+  bool HasAltitudeInfo()
+  {
+    return !m_featureAltitudes.empty();
+  }
+
+  void SortFeatureAltitudes()
+  {
+    sort(m_featureAltitudes.begin(), m_featureAltitudes.end(), my::LessBy(&Processor::TFeatureAltitude::first));
+  }
+
+private:
+  generator::SrtmTileManager m_srtmManager;
+  TFeatureAltitudes m_featureAltitudes;
+  vector<bool> m_altitudeAvailability;
+  TAltitude m_minAltitude;
+};
+
+uint32_t GetFileSize(string const & filePath)
+{
+  uint64_t size;
+  if (!my::GetFileSize(filePath, size))
+  {
+    LOG(LWARNING, ("altitudeAvailability", filePath, "size = 0"));
```

Comment:
my::GetFileSize(filePath, size) returns false in case of exception:
```C++
bool GetFileSize(string const & fName, uint64_t & sz)
{
  try
  {
    typedef my::FileData fdata_t;
    fdata_t f(fName, fdata_t::OP_READ);
    sz = f.Size();
    return true;
  }
  catch (RootException const &)
  {
    // supress all exceptions here
    return false;
  }
}
```
So if (!my::GetFileSize(filePath, size)) then something goes really wrong. 
I need to log this info now to be aware about sizes of parts of altitude section while mwm generation. Later the second log record will be removed.

But I rewritten this method:

```C++
uint32_t GetFileSize(string const & filePath)
{
  uint64_t size;
  if (!my::GetFileSize(filePath, size))
  {
    LOG(LERROR, (filePath, "size = 0"));
    return 0;
  }

  LOG(LINFO, (filePath, "size =", size, "bytes"));
  return size;
}
```